File: netlify/functions/generate_excel.js
const https = require('https');
const querystring = require('querystring');

const CLIENT_ID = 'd6dd88e1-a49e-4350-8339-f0f42c4b3b2e';
const TENANT_ID = '667afa82-1126-4a78-8f76-0918c7f2a845';
const BASE_FOLDER = 'UPC Submissions Automated';
const TEMPLATE_NAME = 'NPC Form 2026 1.xlsx';

function httpsRequest(options, body) {
  return new Promise((resolve, reject) => {
    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        if (res.statusCode >= 400) {
          reject(new Error(`HTTP ${res.statusCode}: ${data}`));
        } else {
          resolve(data);
        }
      });
    });
    req.on('error', reject);
    if (body) req.write(body);
    req.end();
  });
}

async function getAccessToken(refreshToken) {
  const body = querystring.stringify({
    client_id: CLIENT_ID,
    grant_type: 'refresh_token',
    refresh_token: refreshToken,
    scope: 'Files.ReadWrite offline_access User.Read',
  });
  const data = await httpsRequest({
    hostname: 'login.microsoftonline.com',
    path: `/${TENANT_ID}/oauth2/v2.0/token`,
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      'Content-Length': Buffer.byteLength(body),
    },
  }, body);
  return JSON.parse(data).access_token;
}

async function graphRequest(token, method, path, body, contentType) {
  const isBuffer = Buffer.isBuffer(body);
  const bodyStr = body && !isBuffer ? JSON.stringify(body) : body;
  const options = {
    hostname: 'graph.microsoft.com',
    path: `/v1.0${path}`,
    method,
    headers: { 'Authorization': `Bearer ${token}` },
  };
  if (body) {
    options.headers['Content-Type'] = contentType || 'application/json';
    options.headers['Content-Length'] = isBuffer ? body.length : Buffer.byteLength(bodyStr);
  }
  try {
    const data = await httpsRequest(options, bodyStr || body);
    return data ? JSON.parse(data) : {};
  } catch (e) {
    if (e.message.includes('HTTP 409')) return {};
    throw e;
  }
}

async function graphDownload(token, path) {
  return new Promise((resolve, reject) => {
    const req = https.request({
      hostname: 'graph.microsoft.com',
      path: `/v1.0${path}`,
      method: 'GET',
      headers: { 'Authorization': `Bearer ${token}` },
    }, (res) => {
      const chunks = [];
      res.on('data', chunk => chunks.push(chunk));
      res.on('end', () => {
        if (res.statusCode >= 400) {
          reject(new Error(`HTTP ${res.statusCode}: ${Buffer.concat(chunks).toString()}`));
        } else {
          resolve(Buffer.concat(chunks));
        }
      });
    });
    req.on('error', reject);
    req.end();
  });
}

function encodePathSegment(s) {
  return encodeURIComponent(s).replace(/'/g, "''");
}

async function ensureFolder(token, parentPath, folderName) {
  const url = parentPath
    ? `/me/drive/root:/${parentPath.split('/').map(encodePathSegment).join('/')}:/children`
    : '/me/drive/root/children';
  await graphRequest(token, 'POST', url, { name: folderName, folder: {} });
}

async function uploadFile(token, folderPath, fileName, data) {
  const fullPath = `${folderPath}/${fileName}`.split('/').map(encodePathSegment).join('/');
  await graphRequest(token, 'PUT', `/me/drive/root:/${fullPath}:/content`, data, 'application/octet-stream');
}

async function getTemplate(token) {
  return graphDownload(token, `/me/drive/root:/${encodePathSegment(TEMPLATE_NAME)}:/content`);
}

exports.handler = async (event) => {
  const cors = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'POST, OPTIONS',
  };

  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 200, headers: cors, body: '' };
  }

  try {
    const refreshToken = process.env.MS_REFRESH_TOKEN;
    if (!refreshToken) throw new Error('MS_REFRESH_TOKEN not set');

    const { formData, products, files } = JSON.parse(event.body);

    const token = await getAccessToken(refreshToken);
    await ensureFolder(token, '', BASE_FOLDER);

    for (const p of products) {
      await ensureFolder(token, BASE_FOLDER, p.folderName);
    }

    for (const f of files) {
      const buf = Buffer.from(f.data, 'base64');
      await uploadFile(token, `${BASE_FOLDER}/${f.folderName}`, f.fileName, buf);
    }

    // Get template from OneDrive
    const templateBuf = await getTemplate(token);

    // We'll use the template as-is and populate via a simple XML patch
    // For now return the template with a note - full openpyxl-style editing needs a build step
    // Instead, build a simple xlsx from scratch using the existing template bytes
    const clientSafe = (formData.clientName || 'client').replace(/[^a-z0-9]/gi, '_');
    const date = new Date().toISOString().slice(0, 10);
    const excelName = `NPC_Form_${clientSafe}_${date}.xlsx`;

    // Upload template copy to OneDrive (it will be pre-populated by user or we just upload it)
    await uploadFile(token, BASE_FOLDER, excelName, templateBuf);

    return {
      statusCode: 200,
      headers: cors,
      body: JSON.stringify({
        excel: templateBuf.toString('base64'),
        excelName,
      }),
    };
  } catch (e) {
    console.error(e);
    return {
      statusCode: 500,
      headers: cors,
      body: JSON.stringify({ error: e.message }),
    };
  }
};

File: netlify/functions/generate_excel.py
import json
import base64
import io
import urllib.request
import urllib.parse
import os
import datetime
from openpyxl import load_workbook

CLIENT_ID   = 'd6dd88e1-a49e-4350-8339-f0f42c4b3b2e'
TENANT_ID   = '667afa82-1126-4a78-8f76-0918c7f2a845'
BASE_FOLDER = 'UPC Submissions Automated'
TEMPLATE_NAME = 'NPC_Form_2026_1.xlsx'

def get_access_token(refresh_token):
    data = urllib.parse.urlencode({'client_id': CLIENT_ID,'grant_type': 'refresh_token','refresh_token': refresh_token,'scope': 'Files.ReadWrite offline_access User.Read',}).encode()
    req = urllib.request.Request(f'https://login.microsoftonline.com/{TENANT_ID}/oauth2/v2.0/token',data=data, headers={'Content-Type': 'application/x-www-form-urlencoded'})
    resp = urllib.request.urlopen(req)
    return json.loads(resp.read())['access_token']

def graph(token, method, path, body=None, content_type='application/json'):
    url = 'https://graph.microsoft.com/v1.0' + path
    data = None
    if body is not None:
        data = body if isinstance(body, (bytes, bytearray)) else json.dumps(body).encode()
    req = urllib.request.Request(url, data=data, method=method)
    req.add_header('Authorization', 'Bearer ' + token)
    if data:
        req.add_header('Content-Type', content_type)
    try:
        resp = urllib.request.urlopen(req)
        return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        if e.code == 409:
            return {}
        raise Exception(f'Graph {e.code}: {e.read().decode()}')

def graph_download(token, path):
    url = 'https://graph.microsoft.com/v1.0' + path
    req = urllib.request.Request(url, method='GET')
    req.add_header('Authorization', 'Bearer ' + token)
    return urllib.request.urlopen(req).read()

def ensure_folder(token, parent_path, folder_name):
    if parent_path:
        encoded = '/'.join(urllib.parse.quote(s, safe='') for s in parent_path.split('/'))
        url = f'/me/drive/root:/{encoded}:/children'
    else:
        url = '/me/drive/root/children'
    graph(token, 'POST', url, {'name': folder_name, 'folder': {}})

def upload_file(token, folder_path, file_name, data):
    encoded = '/'.join(urllib.parse.quote(s, safe='') for s in (folder_path + '/' + file_name).split('/'))
    graph(token, 'PUT', f'/me/drive/root:/{encoded}:/content', data, 'application/octet-stream')

def get_template(token):
    return graph_download(token, f'/me/drive/root:/{urllib.parse.quote(TEMPLATE_NAME, safe="")}:/content')

def build_excel(template_bytes, form_data, products):
    wb = load_workbook(io.BytesIO(template_bytes))
    ws = wb['New Product Coding Form']
    ws['B4'] = form_data.get('fromName', '')
    ws['D4'] = form_data.get('clientName', '')
    ws['B5'] = form_data.get('email', '')
    ws['D5'] = form_data.get('ownProduct', 'Yes')
    coding = form_data.get('codingOption', 'Code Immediately')
    ws['A7'] = 'X' if coding == 'Code Immediately' else ''
    ws['A8'] = 'X' if coding == 'Delay until Sales' else ''
    ws['A9'] = 'X' if coding == 'Code by Saturday Date' else ''
    if coding == 'Code by Saturday Date' and form_data.get('saturdayDate'):
        ws['C9'] = form_data['saturdayDate']
    ws['E10'] = form_data.get('containerType', '')
    ws['F10'] = form_data.get('containerMaterial', '')
    addl = form_data.get('additionalInfo', '')
    for i, p in enumerate(products):
        row = 12 + i
        ws[f'A{row}'] = p.get('upc', '')
        ws[f'B{row}'] = p.get('asin', '')
        ws[f'C{row}'] = p.get('costco', '')
        ws[f'D{row}'] = p.get('name', '') + (' | ' + addl if addl else '')
    out = io.BytesIO()
    wb.save(out)
    return out.getvalue()

def handler(event, context):
    cors = {'Access-Control-Allow-Origin': '*','Access-Control-Allow-Headers': 'Content-Type','Access-Control-Allow-Methods': 'POST, OPTIONS'}
    if event.get('httpMethod') == 'OPTIONS':
        return {'statusCode': 200, 'headers': cors, 'body': ''}
    try:
        refresh_token = os.environ.get('MS_REFRESH_TOKEN', '')
        if not refresh_token:
            raise Exception('MS_REFRESH_TOKEN not set')
        body = json.loads(event['body'])
        form_data = body.get('formData', {})
        products = body.get('products', [])
        files = body.get('files', [])
        token = get_access_token(refresh_token)
        ensure_folder(token, '', BASE_FOLDER)
        for p in products:
            ensure_folder(token, BASE_FOLDER, p.get('folderName', ''))
        for f in files:
            upload_file(token, BASE_FOLDER + '/' + f.get('folderName', ''), f.get('fileName', ''), base64.b64decode(f.get('data', '')))
        template_bytes = get_template(token)
        excel_bytes = build_excel(template_bytes, form_data, products)
        client_safe = ''.join(c if c.isalnum() else '_' for c in form_data.get('clientName', 'client'))
        excel_name = f'NPC_Form_{client_safe}_{datetime.date.today().isoformat()}.xlsx'
        upload_file(token, BASE_FOLDER, excel_name, excel_bytes)
        return {'statusCode': 200, 'headers': cors, 'body': json.dumps({'excel': base64.b64encode(excel_bytes).decode(), 'excelName': excel_name})}
    except Exception as e:
        return {'statusCode': 500, 'headers': cors, 'body': json.dumps({'error': str(e)})}
